File: eduPlatformBackend/src/main/java/uz/eduplatform/modules/content/domain/Topic.java
package uz.eduplatform.modules.content.domain;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import uz.eduplatform.modules.auth.domain.User;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "topics")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
@org.hibernate.annotations.SQLRestriction("deleted_at IS NULL")
public class Topic {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "subject_id", nullable = false)
    private Subject subject;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "grade_level", nullable = false)
    private Integer gradeLevel;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Topic parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC")
    @Builder.Default
    private List<Topic> children = new ArrayList<>();

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(nullable = false, columnDefinition = "jsonb")
    private Map<String, String> name;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, String> description;

    @Builder.Default
    @Column(nullable = false)
    private Integer level = 1;

    @Column(name = "path")
    private String path;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @Builder.Default
    @Column(name = "question_count")
    private Integer questionCount = 0;

    @Builder.Default
    @Column(name = "sort_order")
    private Integer sortOrder = 0;

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    public static final int MAX_DEPTH = 5;
}

File: eduPlatformFrontend/src/types/topic.ts
export interface TopicDto {
  id: string;
  subjectId: string;
  parentId: string | null;
  userId: string;
  gradeLevel: number;
  name: string;
  description: string | null;
  nameTranslations: Record<string, string> | null;
  descriptionTranslations: Record<string, string> | null;
  level: number;
  path: string;
  isActive: boolean;
  questionCount: number;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface TopicTreeDto {
  id: string;
  subjectId: string;
  parentId: string | null;
  userId: string;
  gradeLevel: number;
  name: string;
  description: string | null;
  nameTranslations: Record<string, string> | null;
  descriptionTranslations: Record<string, string> | null;
  level: number;
  questionCount: number;
  sortOrder: number;
  isActive: boolean;
  children: TopicTreeDto[];
}

export interface CreateTopicRequest {
  name: Record<string, string>;
  description?: Record<string, string>;
  parentId?: string;
  gradeLevel: number;
}

export interface UpdateTopicRequest {
  name?: Record<string, string>;
  description?: Record<string, string>;
  isActive?: boolean;
}

export interface MoveTopicRequest {
  newParentId?: string | null;
  newSubjectId?: string;
}

export interface ReorderTopicsRequest {
  items: { id: string; sortOrder: number }[];
}
